// @screencloud/apps-sdk/lib/constants.js
// Message Types
export var CONNECT = "CONNECT";
export var CONNECT_SUCCESS = "CONNECT_SUCCESS";
export var DISCONNECT = "DISCONNECT"; // Deprecated.
export var INITIALIZE = "initialize";
export var INITIALIZED = "initialized";
export var START = "start";
export var STARTED = "started";
// Message "type"s which prefix their string with ___
export var LEGACY_PREFIXED_TYPES = [CONNECT, CONNECT_SUCCESS, DISCONNECT];
// Prepended to the start of each log message
export var LOG_PREFIX = "[Apps SDK] ";
// Default INITIALIZE payload for local dev and testing.
export var SAMPLE_INITIALIZE_PAYLOAD = {
    appId: "test-app-111",
    appInstanceId: "test-instance-222",
    config: {
        hello: "world",
        meaningOfLife: 42,
    },
    context: {
        userInteractionEnabled: false,
        loggingLevel: 1,
        playerHeight: 1920,
        playerWidth: 1080,
        region: "eu",
        timezone: "Europe/London",
        appViewerToken: "",
    },
    orgId: "org-333",
    spaceId: "space-444",
    screenId: "screen-555",
    device: {},
    filesByAppInstanceId: { nodes: [] },
};
